### PR TITLE
Ch127/allow users to set the allow country list

### PR DESCRIPTION
**Pull request by AlineRibeiro**

### CODE REVIEW

File: db/migrate/20210819215201_create_countries.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateCountries < ActiveRecord::Migration[6.1]
+  def change
+    create_table :countries do |t|
+      t.string :name
+      t.string :code
```
Comment (guialbuk) [inline db/migrate/20210819215201_create_countries.rb]: Please add `null: false` to them as we want to make sure we don't have null values.

You can also add a unique index to `code`

File: app/models/country.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+class Country < ApplicationRecord
+  has_many :country_permissions, dependent: :destroy
+  has_many :videos, through: :country_permissions
```
Comment (guialbuk) [inline app/models/country.rb:5]: Add presence validation to `name` and `code` and uniqueness validation to `code`.

File: db/migrate/20210804052315_add_session_id_to_video_stream_events.rb [hunk 3]
```diff
@@ -2,6 +2,6 @@
 
 class AddSessionIdToVideoStreamEvents < ActiveRecord::Migration[6.1]
   def change
-    add_column :video_stream_events, :session_id, :string
+    add_column :video_stream_events, :session_id, :string, null: false
```
Comment (guialbuk) [inline db/migrate/20210804052315_add_session_id_to_video_stream_events.rb]: Can you please revert this change?

File: db/seeds.rb [hunk 4]
```diff
@@ -28,3 +29,8 @@
   locations << Geolocation.new(row.to_hash)
 end
 Geolocation.import! locations
+
+countries_text = File.read(Rails.root.join('db/countries.json'))
+countries = JSON.parse(countries_text)
+
+countries.each { |key, value| Country.create(code: key, name: value) }
```
Comment (guialbuk) [inline db/seeds.rb]: You can use `import!` the same way you did for `Geolocation` to make it as fast as possible.

File: spec/controllers/videos_controller_spec.rb [hunk 5]
```diff
@@ -46,6 +47,12 @@
 
         expect(AttachSourceFileJob).to have_been_enqueued.exactly(:once)
       end
+
+      it 'creates a new country permission' do
+        expect do
+          valid_request
+        end.to change(CountryPermission, :count).by(1)
```
Comment (guialbuk) [inline spec/controllers/videos_controller_spec.rb:54]: 🎖️  RSpec

File: app/models/video.rb [hunk 6]
```diff
@@ -21,10 +21,13 @@ class Video < ApplicationRecord
...
   has_many :video_stream_events, dependent: :nullify
   has_one_attached :source_file
+  has_many :country_permissions, dependent: :destroy
+  has_many :countries, through: :country_permissions
 
   enum status: { processing: 0, ready: 1, failed: 2 }, _default: :processing
+  enum country_permission_type: { allowed: 0, denied: 1 }
 
-  validates :source, :status, presence: true
+  validates :source, :status, :country_permission_type, presence: true
```
Comment (guialbuk) [inline app/models/video.rb]:
```suggestion
  validates :source, :status, presence: true
```

File: app/controllers/countries_controller.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class CountriesController < ApplicationController
+  def index
```
Comment (guialbuk) [inline app/controllers/countries_controller.rb:6]: This controller needs authentication.

File: app/controllers/videos_controller.rb [hunk 8]
```diff
@@ -45,6 +45,6 @@ def find_video
   end
 
   def video_params
-    params.require(:video).permit(:title, :published, :source)
+    params.require(:video).permit(:title, :published, :source, :country_permission_type, country_ids: [])
```
Comment (guialbuk) [inline app/controllers/videos_controller.rb:48]: Can we change the payload to be able to pass country codes?

File: spec/factories/country_permissions.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :country_permission do
+    country { 1 }
+    video { 1 }
```
Comment (guialbuk) [inline spec/factories/country_permissions.rb]:
```suggestion
    country
    video
```

This way it creates the associated models and does not hardcode an id.

File: spec/factories/countries.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :country do
+    name { 'Canada' }
+    code { 'ca' }
```
Comment (guialbuk) [inline spec/factories/countries.rb]: As`code` must be unique, using Faker should make [generating country codes](https://github.com/faker-ruby/faker/blob/master/doc/default/address.md).

Faker also has an option that ensures it creates unique values within a test: https://github.com/faker-ruby/faker#ensuring-unique-values

File: app/controllers/videos_controller.rb (hunk 8 above)
Comment (AlineRibeiro) [reply app/controllers/videos_controller.rb:48]: I tried it a few different ways, but couldn't make it work.
